### PR TITLE
Cache the buffered extent value

### LINKED ISSUE #3057: ol.render.canvas.*#getBufferedMaxExtent optimization
Every time a flat coordinate array is added to the `ol.render.canvas.Replay`, the buffered max extent (maxextent + line width padding) is computed. See if the result can be cached.

### PR DESCRIPTION
**Pull request by fredj**
Assuming that the maxExtent and resolution property are immutable

fixes #3057

### CODE REVIEW

**Comment by fredj** [comment]
Thanks for any review

File: src/ol/render/canvas/canvasreplay.js [hunk 1]
```diff
@@ -1116,7 +1126,11 @@ ol.render.canvas.LineStringReplay.prototype.setFillStrokeStyle =
   var strokeStyleMiterLimit = strokeStyle.getMiterLimit();
   this.state_.miterLimit = goog.isDef(strokeStyleMiterLimit) ?
       strokeStyleMiterLimit : ol.render.canvas.defaultMiterLimit;
-  this.maxLineWidth = Math.max(this.maxLineWidth, this.state_.lineWidth);
+  if (this.maxLineWidth != this.state_.lineWidth) {
+    // invalid buffered max extent cache
+    this.bufferedMaxExtent_ = null;
+    this.maxLineWidth = Math.max(this.maxLineWidth, this.state_.lineWidth);
+  }
```
Comment (elemoine) [inline src/ol/render/canvas/canvasreplay.js]: How about not using `Math.max` and do this:

```js
  if (this.state_.lineWidth > this.maxLineWidth) {
    this.maxLineWidth = this.state_.lineWidth;
    // invalidate buffered max extent cache
    this.bufferedMaxExtent_ = null;
  }
```

**Comment by elemoine** [comment]
Shouldn't the same be done in the WebGL replay code?

File: src/ol/render/canvas/canvasreplay.js (hunk 1 above)
Comment (fredj) [inline src/ol/render/canvas/canvasreplay.js]: :+1:

**Comment by fredj** [comment]
The WebGL replay code doesn't use a bufferedMaxExtent

**Comment by elemoine** [comment]
> The WebGL replay code doesn't use a bufferedMaxExtent

Right.

LGTM.

**Comment by fredj** [comment]
Thanks for the review

**Pull request merged**